File: Core/src/main/java/com/inverce/mod/events/annotation/EventInfo.java
package com.inverce.mod.events.annotation;

import java.lang.annotation.Retention;
import java.lang.annotation.Target;

import static java.lang.annotation.ElementType.METHOD;
import static java.lang.annotation.RetentionPolicy.RUNTIME;

@Retention(RUNTIME)
@Target({METHOD})
public @interface EventInfo {

    /**
     * Specifies thread on which callbacks should be run
     */
    ThreadPolicy thread() default ThreadPolicy.CallingThread;

    /**
     * Used when returning AsyncFeature, specifies whatever yield only first value or wait for all calls to return
     */
    boolean yieldAll() default false;
}

File: Events/src/main/java/com/inverce/mod/v2/events/EventInfo.kt
package com.inverce.mod.v2.events

@Retention(AnnotationRetention.RUNTIME)
@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER)
annotation class EventInfo(val thread: ThreadPolicy = ThreadPolicy.CallingThread)

enum class ThreadPolicy {
    BgThread, UiThread, CallingThread
}